crates/bevy_ecs/src/schedule/state.rs

```diff
@@ -147,24 +551,42 @@ pub fn run_enter_schedule<S: States>(world: &mut World) {
 /// - Runs the [`OnTransition { from: exited_state, to: entered_state }`](OnTransition), if it exists.
 /// - Runs the [`OnEnter(entered_state)`] schedule, if it exists.
 pub fn apply_state_transition<S: States>(world: &mut World) {
-    // We want to take the `NextState` resource,
-    // but only mark it as changed if it wasn't empty.
-    let mut next_state_resource = world.resource_mut::<NextState<S>>();
-    if let Some(entered) = next_state_resource.bypass_change_detection().0.take() {
-        next_state_resource.set_changed();
-
-        let mut state_resource = world.resource_mut::<State<S>>();
-        if *state_resource != entered {
+    let Some(next_state_resource) = world.get_resource::<NextState<S>>() else {
+        return;
+    };
+    let Some(current_state) = world.get_resource::<State<S>>().map(|s| s.0.clone()) else {
+        return;
+    };
+    let entered = match next_state_resource {
+        NextState::Keep => None,
+        NextState::Value(v) => Some(v.clone()),
+        NextState::Setter(f) => Some(f(current_state.clone())),
+    };
+    if let Some(entered) = entered {
+        world.insert_resource(NextState::<S>::Keep);
+        if current_state != entered {
+            world.insert_resource(PreviousState(current_state));
+            let mut state_resource = world.resource_mut::<State<S>>();
             let exited = mem::replace(&mut state_resource.0, entered.clone());
             // Try to run the schedules if they exist.
             world.try_run_schedule(OnExit(exited.clone())).ok();
+            world
+                .try_run_schedule(OnStateExit::<S>(PhantomData::<S>))
```

Comment:
I think we should use ::default here, rather than the PhantomData type, just to make this easier to read.